imports/lib/schemas/User.ts

```diff
@@ -14,56 +13,35 @@ declare module 'meteor/meteor' {
       displayName?: string;
       googleAccount?: string;
       googleAccountId?: string;
-      discordAccount?: t.TypeOf<typeof DiscordAccount>;
+      discordAccount?: DiscordAccountType;
       phoneNumber?: string;
       dingwords?: string[];
     }
   }
 }
 
-export const UserCodec = t.type({
-  username: t.union([t.string, t.undefined]),
-  emails: t.array(t.type({
-    address: t.string,
-    verified: t.boolean,
-  })),
-  createdAt: date,
-  lastLogin: t.union([date, t.undefined]),
-  services: t.union([t.UnknownRecord, t.undefined]),
-  roles: t.union([t.UnknownRecord, t.undefined]),
-  hunts: t.union([t.array(t.string), t.undefined]),
-  displayName: t.union([t.undefined, t.string]),
-  googleAccount: t.union([t.string, t.undefined]),
-  googleAccountId: t.union([t.string, t.undefined]),
-  discordAccount: t.union([DiscordAccount, t.undefined]),
-  phoneNumber: t.union([t.string, t.undefined]),
-  dingwords: t.union([t.array(t.string), t.undefined]),
+// Note: this needs to exactly match the type of Meteor.User, otherwise we will
+// fail typechecking when we use our attachSchema function. Also, because
+// Meteor.users isn't a Model, we can't rely on transforms or defaults in this
+// schema.
+export const User = z.object({
+  _id: stringId,
+  username: z.string().regex(/^[a-z0-9A-Z_]{3,15}$/).optional(),
+  emails: z.object({ address: z.string().regex(Email), verified: z.boolean() }).array().optional(),
+  createdAt: z.date().optional(),
+  lastLogin: z.date().optional(),
+  services: z.any().optional(),
+  profile: z.object({}).optional(),
+  roles: z.record(z.string(), z.string().array()).optional(),
+  hunts: z.string().regex(Id).array().optional(),
+  displayName: z.string().optional(),
```

Comment:
Do we want these `z.string()`s to be `nonEmptyString`s?

Reply:
Yeah, I think we do. I added a manual call to `validateSchema` here so we can make sure this meets whatever our standards are. (Though if we ever decided to wrap `Meteor.users` into a `Model`, we could drop it)